src/esw/science.py

```diff
@@ -0,0 +1,456 @@
+        """
+        arr = tx_msg.split(",")
+        if len(arr) < 4:
+            return
+        heater_state = [bool(int(arr[1])), bool(int(arr[2])), bool(int(arr[3]))]
+        ros_msg = HeaterData(state=heater_state)
+        self._publisher_by_tag[arr[0][3:]].publish(ros_msg)
+
+    def _heater_transmit(self, device: int, enable: bool) -> bool:
+        """Sends a UART message to the STM32 chip commanding the state of a
+        particular heater device.
+            - Format: <"$HEATER_CMD,<DEVICE>,<ENABLE>">
+        :param device: An int that is the heater device that should be changed.
+        :param enable: A boolean that is the requested heater state of that
+            device.
+        :returns: A boolean that is the success of the transaction.
+        """
+        tx_msg = f"$HEATER_CMD,{device},{int(enable)}"
+        success = self._send_msg(tx_msg)
+
+        return success
+
+    def _read_msg(self) -> str:
+        """Reads a message on the UART receive line from the STM32 chip.
+        :returns: A string that is the received message. Note that this may be
+            an empty string if there are no messages.
+        """
+        try:
+            self._uart_lock.acquire()
+            msg = self.ser.readline()
+            self._uart_lock.release()
+        except serial.SerialException:
+            if self._uart_lock.locked():
+                self._uart_lock.release()
+            rospy.logerr("Errored in _read_msg")
+        return str(msg)
+
+    def _science_thermistor_handler(self, tx_msg: str) -> None:
+        """Processes a UART message that contains data of the three
+        thermistors onboard the carousel in Celsius. We know which sensor
+        is which because we know what the site is.
+        Then publishes to the associated ros channel.
+        :param tx_msg: A string that was received from UART that contains data
+            of the three carousel temperature sensors.
+            - Format: $SCIENCE_TEMP,<TEMP_0>,<TEMP_1>,<TEMP_2>,<EXTRA_PADDING>`
+        """
+        arr = tx_msg.split(",")
+        if len(arr) < 4:
+            return
+
+        temperature_values = ScienceTemperature()
+
+        for i in range(self._num_science_thermistors):
+            temperature_values.temperatures[i] = float(arr[i + 1])
+
+        self._publisher_by_tag[arr[0][3:]].publish(temperature_values)
+
+    def _send_mosfet_msg(self, device_name: str, enable: bool) -> bool:
+        """Sends a MOSFET message on the UART transmit line to the STM32 chip.
+        :param device_name: A string that is the device that wants to be
+            changed.
+        :param enable: A boolean that is the requested device state.
+        :returns: A boolean that is the success of sent UART transaction.
+        """
+        translated_device = self._mosfet_number_by_device_name[device_name]
+        tx_msg = self._format_mosfet_msg(translated_device, enable)
```

Comment:
I'm not usually one to complain about too much abstraction but calling a one-liner from only one spot seems a little unnecessary and kinda makes the code bloated. Is there any benefit to this?

Reply:
Okay maybe it's fine, there could be some other stuff added to this function that I discussed below. However, at the very least it might be nice to move format_mosfet_msg() down here by send_mosfet_msg() since it's only called here